File: SDKSamples/quickstartjava/src/main/java/nanorep/com/quickstartjava/MainActivity.java
package nanorep.com.quickstartjava;

import android.os.Bundle;
import android.util.Log;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;

import com.integration.core.StateEvent;
import com.nanorep.convesationui.structure.controller.ChatController;
import com.nanorep.convesationui.structure.controller.ChatEventListener;
import com.nanorep.convesationui.structure.controller.ChatLoadResponse;
import com.nanorep.convesationui.structure.controller.ChatLoadedListener;
import com.nanorep.nanoengine.AccountInfo;
import com.nanorep.nanoengine.bot.BotAccount;
import com.nanorep.nanoengine.model.configuration.ConversationSettings;
import com.nanorep.sdkcore.utils.NRError;


/**
 * Please follow the related docs: https://logmein-bold-mobile.github.io/bold360-mobile-docs-android/docs/quick_start/
 * The samples app: https://github.com/bold360ai/bold360-mobile-samples-android.git
 */
public class MainActivity extends AppCompatActivity implements ChatEventListener {

    BotAccount botAccount;

    ChatController chatController;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Fill your account credentials here:
        botAccount = new BotAccount("","","","");

        chatController = createChat(botAccount, new ChatLoadedListener(){
            @Override
            public void onComplete(ChatLoadResponse response) {
                if (response.getError() != null) onError(response.getError());
                else {
                    findViewById(R.id.progressBar).setVisibility(View.GONE);
                    Fragment chatFragment = response.getFragment();
                    if (chatFragment == null)
                        onError(new NRError(NRError.EmptyError, "Chat UI failed to init"));
                    else getSupportFragmentManager().beginTransaction().replace(R.id.content_main, chatFragment).commit();
                }
            }
        });
    }

    /**
     *
     * @param botAccount - The wanted account
     * @param chatLoadedListener - a callback of the chat creation
     * @return ChatController
     */
    ChatController createChat(BotAccount botAccount, ChatLoadedListener chatLoadedListener) {

        return new ChatController.Builder(this)
                .conversationSettings(new ConversationSettings()).chatEventListener(this)
                .chatEventListener(this)
                .build(botAccount, chatLoadedListener);
    }


    // Chat events handling by the app:
    @Override
    public void onAccountUpdate(@NonNull AccountInfo accountInfo) {}

    @Override
    public void onChatStateChanged(@NonNull StateEvent stateEvent) {
        Log.d("ChatEventListener","onChatStateChanged");
    }

    @Override
    public void onPhoneNumberSelected(@NonNull String s) {
        Log.d("ChatEventListener","onPhoneNumberSelected");

    }

    @Override
    public void onUploadFileRequest() {
        Log.d("ChatEventListener","onUploadFileRequest");

    }

    @Override
    public void onUrlLinkSelected(@NonNull String s) {
        Log.d("ChatEventListener","onUrlLinkSelected");

    }

    @Override
    public void onError(@NonNull NRError nrError) {
        Log.d("ChatEventListener","onError");
    }

    @Override
    public void onPointerCaptureChanged(boolean hasCapture) {
        Log.d("ChatEventListener","onPointerCaptureChanged");
    }
}

File: SDKSamples/quickstart/src/main/java/nanorep/com/quickstart/MainActivity.kt
package nanorep.com.quickstart

import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import com.nanorep.convesationui.structure.controller.ChatController
import com.nanorep.convesationui.structure.controller.ChatEventListener
import com.nanorep.convesationui.structure.controller.ChatLoadResponse
import com.nanorep.convesationui.structure.controller.ChatLoadedListener
import com.nanorep.convesationui.structure.providers.ChatUIProvider
import com.nanorep.nanoengine.AccountInfo
import com.nanorep.nanoengine.model.configuration.ConversationSettings
import com.nanorep.sdkcore.utils.NRError
import com.nanorep.sdkcore.utils.toast
import nanorep.com.quickstart.databinding.ActivityMainBinding

interface ChatFlowHandler : ChatEventListener {
    fun onAccountReady(account: AccountInfo, chatStartError: (() -> Unit)? = null)
    fun waitingVisibility(visible: Boolean)
}

class MainActivity : AppCompatActivity(), ChatFlowHandler {

    private var _binding: ActivityMainBinding? = null
    private val binding get() = _binding!!

    private var accountProvider = AccountHandler(false)
    private var boldController: ChatController? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        _binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        showFragment(AccountForm(), AccountForm.TAG)
    }

    override fun waitingVisibility(visible: Boolean) {
        binding.progressBar.visibility = if (visible) View.VISIBLE else View.GONE
    }

    override fun onAccountReady(account: AccountInfo, chatStartError: (() -> Unit)?) {
        accountProvider.update(account)
        createChat(account, chatStartError)
    }

    private fun createChat(account: AccountInfo, chatStartError: (() -> Unit)?) {
        waitingVisibility(true)

        boldController = ChatController.Builder(this).apply {

            conversationSettings(ConversationSettings())
            chatEventListener(this@MainActivity)

            // Here we apply the UI provider
            chatUIProvider(coxUIConfig())

        }.build(account, object : ChatLoadedListener {

            override fun onComplete(result: ChatLoadResponse) {

                val error = result.error ?: let {

                    if (result.fragment == null) {
                        NRError(NRError.EmptyError, "Chat UI failed to init")
                    } else {
                        null
                    }
                }

                error?.let {
                    onError(it)
                    chatStartError?.invoke()
                } ?: openConversationFragment(result.fragment!!)

                waitingVisibility(false)
            }
        })
    }

    private fun openConversationFragment(fragment: Fragment) {
        if (isFinishing || supportFragmentManager.isStateSaved ||
            supportFragmentManager.findFragmentByTag(CONVERSATION_FRAGMENT_TAG) != null
        ) {
            return
        }
        showFragment(fragment, CONVERSATION_FRAGMENT_TAG, true)
    }

    override fun onError(error: NRError) {
        super.onError(error)
        toast(this, error.description ?: error.reason ?: error.errorCode, Toast.LENGTH_SHORT)
    }

    private fun showFragment(fragment: Fragment, tag: String, addToBackStack: Boolean = false) {
        supportFragmentManager
            .beginTransaction()
            .replace(R.id.content_main, fragment, tag).apply {
                if (addToBackStack) {
                    addToBackStack(tag)
                }
            }.commit()
    }

    override fun onBackPressed() {
        if (supportFragmentManager.backStackEntryCount == 1) {
            boldController?.destruct()
        }
        super.onBackPressed()
    }

    companion object {
        const val CONVERSATION_FRAGMENT_TAG = "conversation_fragment"
    }


    private fun coxUIConfig() : ChatUIProvider = ChatUIProvider(this).apply {
        chatElementsUIProvider.incomingUIProvider.carouselUIProvider.configure = { adapter ->
            adapter.apply {
                setInfoSubTitleMinLines(4)
                setInfoTitleMinLines(3)
            }
        }
    }
}